### PR TITLE
Feature viewer functionality

### PR DESCRIPTION
**Pull request by HellInsider**

### CODE REVIEW

File: code/allmall.pro.user [hunk 1]
```diff
@@ -0,0 +1,985 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (kirillova-ak5) [inline code/allmall.pro.user]: this file souldnt be commited, add *.pro.user in your gitignore

File: code/core/main.cpp [hunk 2]
```diff
@@ -2,10 +2,11 @@
 
 #include "device\mainwindow.h"
 
+
 int main(int argc, char *argv[])
 {
     QApplication a(argc, argv);
-    main_window w;
+    MainWindow w;
```
Comment (kirillova-ak5) [inline code/core/main.cpp:9]: we change codestyle?

File: code/core/viewer.cpp [hunk 3]
```diff
@@ -1,10 +1,125 @@
...
+        if (image.size() != viewport()->size())
+        {
+            image = QImage(viewport()->size(), QImage::Format_ARGB32_Premultiplied);
+        }
+
+        QPainter imagePainter(&image);
+        QGraphicsView::render(&imagePainter);
+        imagePainter.end();
+
+        QPainter p(viewport());
```
Comment (kirillova-ak5) [inline code/core/viewer.cpp:37]: you draw twise, with imagePainter and with p - why?

File: code/device/viewer.h [hunk 4]
```diff
@@ -1,5 +1,10 @@
 #ifndef VIEWER_H
 #define VIEWER_H
+#include <QGraphicsView>
```
Comment (kirillova-ak5) [inline code/device/viewer.h:3]: includes of big complex classes shoud be in .cpp, in .h use forward declaration

File: code/device/viewer.h [hunk 5]
```diff
@@ -13,20 +18,59 @@ enum viewer_error_code
...
-    void InitMap(svg_view* newMapPic);          // Saving new map imame
+    explicit viewer(QWidget *parent = nullptr);
+    bool InitMap(const QString &fileName);          // Saving new map imame
     void ClearMap();                            // deleting img ptr
 
     void ViewMap();                             // Draws map
     void ViewObject();                          // Drows something (preparing for future) f.e. selected shop
     void ViewGraph();                           // Shows graph on map (for debug )
     void ViewPath(coord *from, coord *to);      // Shows path from firt point to second point (calls Graph)
+    qreal zoomFactor();
```
Comment (kirillova-ak5) [inline code/device/viewer.h]: why qreal instead of standart float and when we change codestyle

File: code/device/viewer.h [hunk 6]
```diff
@@ -13,20 +18,59 @@ enum viewer_error_code
...
+    qreal GetMapPicScale();
```
Comment (kirillova-ak5) [inline code/device/viewer.h]: why qreal

File: code/device/viewer.h [hunk 7]
```diff
@@ -13,20 +18,59 @@ enum viewer_error_code
...
+    virtual ~viewer() {};
```
Comment (kirillova-ak5) [inline code/device/viewer.h]: memory leak at mapScene, must delete it here, and delete all objects created in init with "new"

File: code/device/mainwindow.h [hunk 8]
```diff
@@ -1,32 +1,61 @@
...
     Q_OBJECT
 
 public:
-    main_window();
-    ~main_window();
-    virtual void Render(QPainter *painter);
+    MainWindow();
+    ~MainWindow();
+
+    bool loadFile(const QString &path);
```
Comment (kirillova-ak5) [inline code/device/mainwindow.h]: why we have 2 load file whith different names

File: code/device/mainwindow.h [hunk 9]
```diff
@@ -1,32 +1,61 @@
...
     bool LoadFile(const QString &svgFileName, const QString &xmlFileName);
+    virtual void Render(QPainter *painter);
 
 public slots:
-    void Show();
-    void setNewView(svg_view * toSet);
-
-protected:
+     //void openFile();
```
Comment (kirillova-ak5) [inline code/device/mainwindow.h]: not comment code, delete it

File: code/device/mainwindow.h [hunk 10]
```diff
@@ -1,32 +1,61 @@
...
+   // void exportImage();
     bool event(QEvent *event) override;
     void timerEvent(QTimerEvent *) override;
+    //void setRenderer(int renderMode);
```
Comment (kirillova-ak5) [inline code/device/mainwindow.h]: delete

File: code/device/mainwindow.h [hunk 11]
```diff
@@ -1,32 +1,61 @@
...
+    void Show();
+    void setNewView(QGraphicsSvgItem * toSet);
```
Comment (kirillova-ak5) [inline code/device/mainwindow.h:34]: codestyle

File: code/device/mainwindow.h [hunk 12]
```diff
@@ -1,32 +1,61 @@
...
+    QAction *m_nativeAction;
+    QAction *m_glAction;
+    QAction *m_imageAction;
+    QAction *m_antialiasingAction;
+    QAction *m_backgroundAction;
+    QAction *m_outlineAction;
+
+    QLabel *m_zoomLabel;
 
 private:
```
Comment (kirillova-ak5) [inline code/device/mainwindow.h:49]: delete this "private" if you already have one before

File: code/device/mainwindow.h [hunk 13]
```diff
@@ -1,32 +1,61 @@
...
     void timerEvent(QTimerEvent *) override;
+    //void setRenderer(int renderMode);
+    void Show();
+    void setNewView(QGraphicsSvgItem * toSet);
+
+private slots:
+    void updateZoomLabel();
+
+private:
+    QAction *m_nativeAction;
```
Comment (kirillova-ak5) [inline code/device/mainwindow.h:40]: why are you yse m_ prefix

File: code/device/mainwindow.cpp [hunk 14]
```diff
@@ -31,36 +32,42 @@ void main_window::timerEvent(QTimerEvent *event)
     }
 }
 
-void main_window::setNewView(svg_view * toSet)
+void MainWindow::setNewView(QGraphicsSvgItem *toSet)
 {
-    view = toSet;
+    //mapViewer->InitMap(toSet);
```
Comment (kirillova-ak5) [inline code/device/mainwindow.cpp]: dont commit commented code

File: code/device/mainwindow.cpp [hunk 15]
```diff
@@ -31,36 +32,42 @@ void main_window::timerEvent(QTimerEvent *event)
...
 }
 
-bool main_window::LoadFile(const QString &svgFileName, const QString &xmlFileName)
+bool MainWindow::LoadFile(const QString &svgFileName, const QString &xmlFileName)
 {
     if (QFileInfo::exists(svgFileName) && QFileInfo::exists(xmlFileName)) // maybe move checks to view module? - KYG
     {
         mapInfo->SetAnotherMap(svgFileName, xmlFileName);
```
Comment (kirillova-ak5) [inline code/device/mainwindow.cpp:42]: may be separate check of svg existing and init from check of xml existing and init

File: code/device/svgview.h [hunk 16]
```diff
@@ -25,19 +25,23 @@ class svg_view : public QGraphicsView
 
     QSize SvgSize() const;
     QSvgRenderer *Renderer() const;
-
```
Comment (kirillova-ak5) [inline code/device/svgview.h:28]: should this file be delete?

File: code/device/svgview.cpp [hunk 17]
```diff
@@ -50,7 +50,7 @@ bool svg_view::OpenFile(const QString &fileName)
     QGraphicsSvgItem *black = new QGraphicsSvgItem();
 
     black->setSharedRenderer(svgRenderer);
-    black->setElementId(QLatin1String("shop1_1"));
```
Comment (kirillova-ak5) [inline code/device/svgview.cpp:53]: delete this file

File: code/device/map.cpp [hunk 18]
```diff
@@ -1,21 +1,62 @@
...
+QGraphicsSvgItem* map::GetPic()
+{
+    return pic;
+}
+
+bool map::OpenFile(const QString &fileName)
+{
+    pic = new QGraphicsSvgItem(fileName);
+
+/*
```
Comment (kirillova-ak5) [inline code/device/map.cpp]: fucking fuck delete commented code

File: code/device/map.cpp [hunk 19]
```diff
@@ -26,7 +67,7 @@ graph* CreateGraphFromSvg (svg_view* v)
 
 
 map::map(QObject *parent) :
-    QObject(parent), svgMapFileName(""), pic(new svg_view), mapInfoFileName(""), object_indexes()
+    QObject(parent), svgMapFileName("")/*, pic(new QGraphicsSvgItem)*/, mapInfoFileName(""), object_indexes()
```
Comment (kirillova-ak5) [inline code/device/map.cpp]: delete

File: code/device/map.cpp [hunk 20]
```diff
@@ -64,13 +105,14 @@ void map::SetAnotherMap(QString const & mapToSet, QString const & extrasToSet)
     if ((flags & __MAP_PIC_CHANGED) != 0)
     {
         //ClearPic();
-        pic->OpenFile(mapToSet);
+        //pic->OpenFile(mapToSet);
```
Comment (kirillova-ak5) [inline code/device/map.cpp]: delete

File: code/device/map.cpp [hunk 21]
```diff
@@ -26,7 +67,7 @@ graph* CreateGraphFromSvg (svg_view* v)
 
 
 map::map(QObject *parent) :
-    QObject(parent), svgMapFileName(""), pic(new svg_view), mapInfoFileName(""), object_indexes()
+    QObject(parent), svgMapFileName("")/*, pic(new QGraphicsSvgItem)*/, mapInfoFileName(""), object_indexes()
 {
     //pic = nullptr;
     paths = nullptr;
```
Comment (kirillova-ak5) [inline code/device/map.cpp]: move to initializer list

File: code/device/map.cpp [hunk 22]
```diff
@@ -1,21 +1,62 @@
...
 #define __MAP_ALL_CHANGED ((__MAP_PIC_CHANGED)|(__MAP_GRAPH_CHANGED)|(__MAP_INFO_CHANGED))
 
+QGraphicsSvgItem* map::GetPic()
+{
+    return pic;
+}
+
+bool map::OpenFile(const QString &fileName)
+{
+    pic = new QGraphicsSvgItem(fileName);
```
Comment (kirillova-ak5) [inline code/device/map.cpp:19]: where is "delete" for this "new"

File: graph.cpp [hunk 23]
```diff
@@ -0,0 +1,156 @@
+#include "device\graph.h"
```
Comment (kirillova-ak5) [inline graph.cpp]: why are you create file in root of project?

File: graph.h [hunk 24]
```diff
@@ -0,0 +1,31 @@
+#ifndef GRAPH_H
```
Comment (kirillova-ak5) [inline graph.h]: why are you create file in root of project?

File: main.cpp [hunk 25]
```diff
@@ -0,0 +1,13 @@
+#include <QApplication>
```
Comment (kirillova-ak5) [inline main.cpp]: why are you create file in root of project?

File: mainwindow.cpp [hunk 26]
```diff
@@ -0,0 +1,83 @@
+#include <QtWidgets>
```
Comment (kirillova-ak5) [inline mainwindow.cpp]: why are you create file in root of project?

File: mainwindow.h [hunk 27]
```diff
@@ -0,0 +1,61 @@
+#ifndef MAINWINDOW_H
```
Comment (kirillova-ak5) [inline mainwindow.h]: why are you create file in root of project?

File: mainwindow.ui [hunk 28]
```diff
@@ -0,0 +1,33 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (kirillova-ak5) [inline mainwindow.ui]: why are you create file in root of project?

File: map.cpp [hunk 29]
```diff
@@ -0,0 +1,168 @@
+#include <QSvgRenderer>
```
Comment (kirillova-ak5) [inline map.cpp]: why are you create file in root of project?

File: map.h [hunk 30]
```diff
@@ -0,0 +1,78 @@
+#ifndef MAP_H
```
Comment (kirillova-ak5) [inline map.h]: why are you create file in root of project? and all next files

**Review by kirillova-ak5 — CHANGES_REQUESTED** [review]

File: code/core/main.cpp (hunk 2 above)
Comment (HellInsider) [reply code/core/main.cpp:9]: No, thats ugly first-writed code

**Pull request merged**